drivers/serial/uart_rpi_pico.c

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2021, Yonatan Schachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/uart.h>
+#include <hardware/uart.h>
+#include <hardware/gpio.h>
+
+struct uart_rpi_config {
+	uart_inst_t *const uart_dev;
+	uint32_t baudrate;
+	uint32_t tx_pin;
+	uint32_t rx_pin;
+};
+
+static int uart_rpi_poll_in(const struct device *dev, unsigned char *c)
+{
+	const struct uart_rpi_config *config = dev->config;
+
+	if (!uart_is_readable(config->uart_dev)) {
+		return -1;
+	}
+
+	*c = (unsigned char)uart_get_hw(config->uart_dev)->dr;
+	return 0;
+}
+
+static void uart_rpi_poll_out(const struct device *dev, unsigned char c)
+{
+	const struct uart_rpi_config *config = dev->config;
+
+	uart_putc_raw(config->uart_dev, c);
+}
+
+static int uart_rpi_init(const struct device *dev)
+{
+	const struct uart_rpi_config *config = dev->config;
+
+	gpio_init(config->tx_pin);
+	gpio_init(config->rx_pin);
+	gpio_set_function(config->tx_pin, GPIO_FUNC_UART);
+	gpio_set_function(config->rx_pin, GPIO_FUNC_UART);
```

Comment:
a `pinctrl` driver should be created to handle this. You can refer to https://docs.zephyrproject.org/latest/guides/pinctrl/index.html for more details.

Reply:
Hi @gmarull , I've just pushed a first draft of the pinctrl driver. The HAL part can be found here: https://github.com/yonsch/hal_rp_demo/commit/c62c5a1ba579eb0b3a7f14abe303f6bfe1f7a16b .
It's not ready for submission yet (no documentation, no pinctrl generation), but I would appreciate it if you could take a quick glance just to tell me that I'm in the right direction, before I go on finalizing it. I only had the gigadevice driver as a reference, so I won't be surprised if many things fell out of place.